File: hackerrank/problem solving/algorithms/2 - implementation/8 - migratory birds/migratory_birds.py
# https://www.hackerrank.com/challenges/migratory-birds/problem?isFullScreen=true

from collections import Counter


def main():
    birds_to_counts = read_input()
    print(find_most_spotted_bird(birds_to_counts))


# n: length of input
# T: O(n)
# S: O(n) extra space
def read_input():
    input()
    return Counter(map(int, input().split()))


# n: length of birds_to_counts
# T: O(n)
# S: O(1) extra space
def find_most_spotted_bird(birds_to_counts):
    bird_id, _ = max(birds_to_counts.items(), key=lambda x: (x[1], -x[0]))
    return bird_id


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/8 - migratory birds/migratory_birds.rs
// https://www.hackerrank.com/challenges/migratory-birds/problem?isFullScreen=true

use std::collections::HashMap;
use text_io::read;

fn main() {
    let birds_to_counts = read_input();
    println!("{}", find_most_spotted_bird(&birds_to_counts));
}

// n: length of input
// T: O(n)
// S: O(n) extra space
fn read_input() -> HashMap<i32, i32> {
    let n: usize = read!();
    let mut freq_map = HashMap::new();
    for _ in 0..n {
        let x: i32 = read!();
        *freq_map.entry(x).or_insert(0) += 1;
    }
    freq_map
}

// n: length of birds_to_counts
// T: O(n)
// S: O(1) extra space
fn find_most_spotted_bird(birds_to_counts: &HashMap<i32, i32>) -> i32 {
    *birds_to_counts
        .iter()
        .max_by_key(|(&id, &count)| (count, -id))
        .unwrap()
        .0
}
